### PR TITLE
Streamline cluster install

### PR DESCRIPTION
**Pull request by nicholastmosher**
This adds better error messages to the `helm` client in such a way that it bubbles all the way up to the CLI. Before, if `helm` was not installed, if the user ran `fluvio cluster install`, they would be given the following error:

```bash
$ fluvio cluster install
error: No such file or directory (os error 2)
```

Now, it gives this error:

```bash
$ fluvio cluster install
error: Unable to find 'helm' executable
  Please make sure helm is installed and in your PATH.
  See https://helm.sh/docs/intro/install/ for more help
```

In addition, this PR makes the cluster installer attempt to fix some things if the precheck runs into errors:

- If fluvio-sys is not installed, it attempts to install it
- If the minikube context is not set right, it will set it

### CODE REVIEW

**Comment by nicholastmosher** [comment]
This closes #235

**Comment by nicholastmosher** [comment]
Hold up on merging this, I'm going to add some tracing in here to help with debugging.

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:556]: The point of this is that pre-install can detect things that have exactly one fix. If we can detect a fix and we know exactly what needs to be done, we should just do it.

Comment (sehz) [reply src/cluster/src/install.rs:566]: I agree but we should consolidate.  let's move action to `pre_install_check` and have option or flag to take action.

**Review by sehz — CHANGES_REQUESTED** [review]

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:501]: why use of `block_on` since this is not async fn

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:519]: hmm. since this seems not intuitive, having build method on top of build method.

File: src/cluster/src/lib.rs
Comment (sehz) [inline src/cluster/src/lib.rs:14]: unneeded use of block_on

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:505]: It is weird that ClusterInstaller has dependencies to ClusterInstallBuilder.   It should be other way around?

**Review by sehz — CHANGES_REQUESTED** [review]

**Comment by sehz** [comment]
may need feed back from @Ryanmtate and @nacardin regarding new API

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:501]: `ClusterInstaller::install_fluvio` is an async function, so it needs the `block_on` to drive it

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:519]: This is probably just a problem with the example. There are now two ways to get an installer:

- With all default settings using `ClusterInstaller::new()?`
- With overridden settings using the builder

When using the builder style, you _create_ a builder using `ClusterInstaller::builder()`, and you _finalize_ a builder using `ClusterInstallerBuilder::build()`. So normally you'd have the settings in between those methods, like

```rust
let installer = ClusterInstaller::builder()
    .with_namespace("fluvio")
    .with_rust_log("debug")
    // .with_other_settings()
    .build()?;
```

I'll update the example to make more sense

File: src/cluster/src/lib.rs
Comment (nicholastmosher) [reply src/cluster/src/lib.rs:14]: `install_fluvio` is async

Comment (sehz) [reply src/cluster/src/lib.rs:14]: is this example or dockets?  if it is example, there is no need block on.   it should be just express as async snippet

File: src/cli/src/cluster/install/k8.rs
Comment (sehz) [inline src/cli/src/cluster/install/k8.rs:13]: Typically, in the builder pattern, builder is references not the struct it is being [constructured](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html).

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:640]: Moving pre-check logic into installer will make Installer more complex.

**Review by sehz — CHANGES_REQUESTED** [review]

**Comment by vijaylaxmid** [comment]
I think this also fixes https://github.com/infinyon/fluvio/issues/313

File: src/cli/src/cluster/install/k8.rs
Comment (nicholastmosher) [reply src/cli/src/cluster/install/k8.rs:13]: A few reasons I did it this way:

- The `ClusterInstallerBuilder` acts just like a config object at the end of the day, with all the options neatly laid out. I even call it `config` inside the `ClusterInstaller`
- An API user can avoid importing the `ClusterInstallerBuilder` type because all of the necessary methods are on the `ClusterInstaller` type. E.g.

```rust
use fluvio_cluster::ClusterInstaller;
let installer = ClusterInstaller::builder()
    .with_namespace("my-namespace")
    .with_rust_log("debug")
    .build()?;
installer.install_fluvio().await?;
```

- The `build()` method on the `ClusterInstallerBuilder` is the gate where things either initialize successfully or not. For example, we can do validation on whether any options can conflict with each other. Currently, this would be the point of failure if the HelmClient or K8Client could not be initialized.
- The Builder is using the consuming calls, i.e. they take `mut self`. I didn't want to have tons of consuming calls on the same object that was actually performing the installation.
- After calling `.build()`, all of the settings are locked in. That means you couldn't accidentally call `install_sys`, then change a setting, then call `install_app` with incompatible settings.

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:663]: Where do you imagine it all living? I feel like this crate is definitely the place for it. I would consider the `check` module but the `ClusterInstaller` has all of the parameters that you want the check to be compatible with, so for that reason it also kind of makes sense to put it here.

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:528]: I think sometimes it is laid out that way, but something I like about this pattern is that the `ClusterInstallerBuilder` is pretty much nothing more than a config object that is put together with builder methods where you can do validation along the way. Also, by using the `ClusterInstaller::builder() -> ClusterInstallerBuilder` method, users don't actually need to import the `ClusterInstallerBuilder` type.

File: src/cli/src/cluster/install/k8.rs
Comment (sehz) [reply src/cli/src/cluster/install/k8.rs:13]: provisioning  helm client and K8client should be done at the time of install and should be defer to pre-check module.  Probably pre-check should can return resource as well.

File: src/cluster/src/install.rs
Comment (sehz) [reply src/cluster/src/install.rs:663]: see previous comments.  Can make new struct or module, ResourceDependency.  It will be responsible for checking resources exists and return it if needed it.  That way installer can be simpler

File: src/cluster/src/lib.rs
Comment (nicholastmosher) [reply src/cluster/src/lib.rs:14]: This is a doc test that is marked as `no_run`. So it will do a compile-check, but not run anything. I included the `async_std` line so that even users who don't really know much about async will know that you need to do that in order to get this to run.

File: src/cli/src/cluster/install/k8.rs
Comment (nicholastmosher) [reply src/cli/src/cluster/install/k8.rs:13]: > provisioning helm client and K8client should be done at the time of install

Why is that a requirement? This is ensuring the same order or operations

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:663]: Is that something that we need to do right now? Can we make an issue for that if we're going to start to doing such a big refactor? I don't really think that's super necessary to do immediately

Comment (sehz) [reply src/cluster/src/install.rs:663]: This PR is major API changes. so can defer this and make smallest change now?

Comment (sehz) [reply src/cluster/src/install.rs:663]: Since we want to start publishing crates soon, changes should be limited to additional or add minor behavior changes

**Comment by nicholastmosher** [comment]
I've split out the builder changes and I'll open a new PR with it for further discussion 👍

**Comment by sehz** [comment]
can you rebase from master?

File: src/cluster/src/error.rs
Comment (sehz) [inline src/cluster/src/error.rs:4]: can you split out helm changes only?  this will make easier to digest changes

**Review by sehz — CHANGES_REQUESTED** [review]

File: src/cluster/src/error.rs
Comment (nicholastmosher) [reply src/cluster/src/error.rs:4]: Is there anything you'd like me to change about one or the other? If you're planning on merging them both right away can we just merge them both here? Otherwise I will spend a lot of time trying to extract the changes

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:625]: I kind of feel the opposite way actually. I feel like if there are major changes to be made they should be made before the release when they would not count as a breaking change, rather than after the release when they _would_ be a breaking change

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:73]: chart related option seems to be should be separate out from here.    There are many different options here and this make it harder to incremental add new options later

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:255]: shouldn't this be refer to ClusterInstallBuilder?

File: src/cluster/src/error.rs
Comment (sehz) [reply src/cluster/src/error.rs:4]: add more comments.    There are many question that need to be answered.  The helm changes are straightforward.

File: src/cluster/src/install.rs
Comment (sehz) [reply src/cluster/src/install.rs:625]: major changes takes time to discuss review.   the change doesn't seem to be must have for this release

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:65]: The reason I added that was because when you install a local chart, you need to give a path to the exact chart file. I ran into a bug where I would try `fluvio cluster install --develop` and now that it also tries to install the -sys chart at the same time, it was attempting to install `fluvio-app` from `./k8-util/helm/fluvio-app` and it was also attempting to install `fluvio-sys` from `./k8-util/helm/fluvio-app`. So when we are using remote repositories, they will have the same value, but we want to have the flexibility to give two separate locations when using local directories.

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:73]: What specifically do you mean about chart related options? Right now the entire installation process is pretty much about configuring and installing charts in the correct way. How to install any particular chart is the responsibility of the `HelmClient`, but I think that installing the charts in such a way that works for Fluivo is exactly what the `ClusterInstaller` is meant to do.

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:255]: The way this is set up is that the user doesn't ever need to interact with the `ClusterInstallerBuilder` type. Currently, the `ClusterInstaller::new()` function actually returns a `ClusterInstallerBuilder`, which is why you see this in the documentation. I introduced #339 in order to make things a little bit nicer by making the function names more accurately represent what is happening. If we merge #339, this line would be rewritten as `let installer = ClusterInstaller::builder()...`

**Comment by nicholastmosher** [comment]
I split out a PR for just the helm changes in #342

**Comment by nicholastmosher** [comment]
This PR now contains just the changes for automatically fixing installation errors, building upon the new helm branch in #342

File: src/cluster/src/install.rs
Comment (sehz) [reply src/cluster/src/install.rs:65]: can you past example of error? not able to see use case clearly

File: src/cluster/src/install.rs
Comment (sehz) [reply src/cluster/src/install.rs:73]: if this flag is enabled,  `sys chart` should be installed only if it doesn't exist.

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:75]: can we group both `install_sys` and `update_context` into separate option?  likely, there would be more option so we can be forward compatible

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:99]: This should be refer to ClusterInstallBuilder not ClusterInstaller

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:625]: let's move this under `pre_install_check` so we consolidate it `check and fix` features

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:633]: we should ask from confirmation from user since this is not typical installation behavior for kubernetes

**Review by sehz — CHANGES_REQUESTED** [review]

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:633]: This is gated behind the `update_context` configuration option. If you like, we can make that option default to `false` and enable it with a CLI flag as necessary

Comment (sehz) [reply src/cluster/src/install.rs:633]: sure

**Comment by nicholastmosher** [comment]
Factored the `pre_install_check` function into a bunch of `check_X` functions and moved them to the `check` module. We also now have a dedicated `CheckError` type, which consists of errors we may see specifically when performing the pre-check. Some of the CheckError variants may be automatically fixed.

New pre-check flow:

- `install_fluvio` is invoked
- `install_fluvio` invokes `pre_install`
- `pre_install` executes a loop that does the following:
  - Runs `pre_install_check`, which returns a result.
  - If the result is Ok, return a success from pre_install
  - If the result is Err, pass that error to `pre_install_fix`
  - `pre_install_fix` will match on the error type to see if it can be automatically fixed or not.
    - If the error can be fixed, fixes it and returns Ok
    - If the error cannot be fixed automatically, returns that same error
  - If `pre_install_fix` returned Ok, re-enter the loop to check for subsequent pre-install checks and fixes
  - If `pre_install_fix` returned Err, bubble up to report to the user

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:77]: I'm not sure what you mean by grouping them into a separate option. They are two different configurations.
- `install_sys` causes the installer to check if `fluvio-sys` is installed, and to install it automatically if not
- `update_context` causes the installer to update the `kubectl` context and, if necessary, the /etc/hosts entry

File: src/cluster/src/install.rs
Comment (nicholastmosher) [reply src/cluster/src/install.rs:67]: When I remove the `chart_location_sys` option and I run install with `--develop`, I get the following:

```
$ cargo run --bin fluvio -- cluster install --develop
>> helm version
>> helm list --filter ^fluvio-sys$ --output json
>> helm install fluvio-sys ./k8-util/helm/fluvio-app --namespace default --devel --set cloud=minikube
NAME: fluvio-sys
LAST DEPLOYED: Sat Oct  3 23:13:00 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
>> helm list --filter ^fluvio-sys$ --output json
>> helm list --filter ^fluvio$ --output json
>> helm install fluvio ./k8-util/helm/fluvio-app --namespace default --devel --set image.registry=infinyon --set image.tag=f9c3749690b3a4b0882463baffc0c83f0cd32448 --set cloud=minikube --version 0.6.0-alpha.1
Exited with status code: 1
```

Note the first helm install is incorrectly using the `fluvio-app` local chart location for sys: 
```
helm install fluvio-sys ./k8-util/helm/fluvio-app --namespace default --devel --set cloud=minikube
```

Comment (nicholastmosher) [reply src/cluster/src/install.rs:67]: In develop mode, should the installer use the local system chart at `k8-util/helm/fluvio-sys`? I assume so, I can update that

**Comment by nicholastmosher** [comment]
Now when using the `--develop` flag, both `fluivio-app` and `fluvio-sys` will be loaded from the local chart paths correctly.

```
    Finished dev [unoptimized + debuginfo] target(s) in 0.24s
     Running `target/debug/fluvio cluster install --develop`
>> helm version
>> helm list --filter ^fluvio-sys$ --output json
>> helm install fluvio-sys ./k8-util/helm/fluvio-sys --namespace default --devel --set cloud=minikube
NAME: fluvio-sys
LAST DEPLOYED: Sat Oct  3 23:31:16 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
>> helm list --filter ^fluvio-sys$ --output json
>> helm list --filter ^fluvio$ --output json
>> helm install fluvio ./k8-util/helm/fluvio-app --namespace default --devel --set image.registry=infinyon --set image.tag=f9c3749690b3a4b0882463baffc0c83f0cd32448 --set cloud=minikube --version 0.6.0-alpha.1
NAME: fluvio
LAST DEPLOYED: Sat Oct  3 23:31:17 2020
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
waiting for spu to be provisioned
1 spus provisioned
```

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:717]: should be using `sys` chart name constant

File: src/cluster/src/install.rs
Comment (sehz) [inline src/cluster/src/install.rs:803]: use `app` chart constants

**Review by sehz — CHANGES_REQUESTED** [review]

**Comment by nicholastmosher** [comment]
I've removed the `chart_location_sys` option now that we're appending to a base chart_location.
Can we defer the chart factorizing to another issue? That seems more like a refactoring that is orthogonal to this PR.

**Comment by sehz** [comment]
I agree.  We can postpone changes

**Review by sehz — APPROVED** [review]

**Review by sehz — CHANGES_REQUESTED** [review]
need to bump up version

**Comment by nicholastmosher** [comment]
You got it

**Comment by sehz** [comment]
Sorry, I meant Cluster crate version

**Comment by nicholastmosher** [comment]
This will also close #351

**Review by sehz — APPROVED** [review]

**Pull request merged**
